Add cloud settings for ES output

ChrsMark opened the pull request:
This PR adds `cloud_id` and `cloud_auth` settings so as to provide support for Fluent-bit shipping to ES cluster in Elastic cloud with an easily configurable manner.

Addresses https://github.com/fluent/fluent-bit/issues/2833

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [x] Debug log output from testing the change
<!-- Invoke Fluent Bit and Valgrind as: $ valgrind ./bin/fluent-bit <args> -->
- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [x] Documentation required for this feature (https://github.com/fluent/fluent-bit-docs/pull/430)

<!--  Doc PR (not required but highly recommended) -->

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

ChrsMark commented:
### Example configuration file for the change
```
INPUT]
  Name forward

[OUTPUT]
  Name es
  Include_Tag_Key true
  Tag_Key tags
  tls On
  tls.verify Off
  cloud_id elastic-obs-deployment:ZXVybxxxxxxxxxxxg==
  cloud_auth elastic:2vxxxxxxxxYV
```

### Debug log output from testing the change
```
./bin/fluent-bit -c ~/fluentbit/pipelines/minimal.conf -v                                                                                   ✔  10370  17:56:49
Fluent Bit v1.7.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/12/07 17:56:52] [ info] Configuration:
[2020/12/07 17:56:52] [ info]  flush time     | 5.000000 seconds
[2020/12/07 17:56:52] [ info]  grace          | 5 seconds
[2020/12/07 17:56:52] [ info]  daemon         | 0
[2020/12/07 17:56:52] [ info] ___________
[2020/12/07 17:56:52] [ info]  inputs:
[2020/12/07 17:56:52] [ info]      forward
[2020/12/07 17:56:52] [ info] ___________
[2020/12/07 17:56:52] [ info]  filters:
[2020/12/07 17:56:52] [ info] ___________
[2020/12/07 17:56:52] [ info]  outputs:
[2020/12/07 17:56:52] [ info]      es.0
[2020/12/07 17:56:52] [ info] ___________
[2020/12/07 17:56:52] [ info]  collectors:
[2020/12/07 17:56:52] [ info] [engine] started (pid=76030)
[2020/12/07 17:56:52] [debug] [engine] coroutine stack size: 12288 bytes (12.0K)
[2020/12/07 17:56:52] [debug] [storage] [cio stream] new stream registered: forward.0
[2020/12/07 17:56:52] [ info] [storage] version=1.1.0, initializing...
[2020/12/07 17:56:52] [ info] [storage] in-memory
[2020/12/07 17:56:52] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/12/07 17:56:52] [debug] [in_fw] Listen='0.0.0.0' TCP_Port=24224
[2020/12/07 17:56:52] [ info] [input:forward:forward.0] listening on 0.0.0.0:24224
[2020/12/07 17:56:52] [debug] [es:es.0] created event channels: read=22 write=23
[2020/12/07 17:56:52] [debug] [output:es:es.0] host=ba133372c73b4721a1f45c73ba885114.europe-west1.gcp.cloud.es.io port=443 uri=/_bulk index=fluent-bit type=_doc
[2020/12/07 17:56:52] [debug] [router] default match rule forward.0:es.0
[2020/12/07 17:56:52] [ info] [sp] stream processor started


[2020/12/07 17:56:57] [debug] [task] created task=0x7fe014c000d0 id=0 OK
[2020/12/07 17:56:57] [debug] [task] created task=0x7fe014c00170 id=1 OK
[2020/12/07 17:56:57] [debug] [task] created task=0x7fe014c00210 id=2 OK
[2020/12/07 17:56:57] [debug] [http_client] not using http_proxy for header
[2020/12/07 17:56:57] [debug] [http_client] header=POST /_bulk HTTP/1.1
Host: ba133372c73b4721a1f45c73ba885114.europe-west1.gcp.cloud.es.io
Content-Length: 2756833
Content-Type: application/x-ndjson
Authorization: Basic ZWxhc3RpYzoydlJjZmZzelNrVXI0eWlxUTJPOEVRWVY=
User-Agent: Fluent-Bit


[2020/12/07 17:56:57] [debug] [http_client] not using http_proxy for header
[2020/12/07 17:56:57] [debug] [http_client] header=POST /_bulk HTTP/1.1
Host: ba133372c73b4721a1f45c73ba885114.europe-west1.gcp.cloud.es.io
Content-Length: 2782242
Content-Type: application/x-ndjson
Authorization: Basic ZWxhc3RpYzoydlJjZmZzelNrVXI0eWlxUTJPOEVRWVY=
User-Agent: Fluent-Bit


[2020/12/07 17:56:57] [debug] [http_client] not using http_proxy for header
[2020/12/07 17:56:57] [debug] [http_client] header=POST /_bulk HTTP/1.1
Host: ba133372c73b4721a1f45c73ba885114.europe-west1.gcp.cloud.es.io
Content-Length: 134364
Content-Type: application/x-ndjson
Authorization: Basic ZWxhc3RpYzoydlJjZmZzelNrVXI0eWlxUTJPOEVRWVY=
User-Agent: Fluent-Bit


[2020/12/07 17:57:02] [debug] [task] created task=0x7fe021c20410 id=3 OK
[2020/12/07 17:57:02] [debug] [task] created task=0x7fe021c20720 id=4 OK
[2020/12/07 17:57:02] [debug] [task] created task=0x7fe021c207e0 id=5 OK
[2020/12/07 17:57:02] [debug] [task] created task=0x7fe021c208a0 id=6 OK
[2020/12/07 17:57:02] [debug] [output:es:es.0] HTTP Status=200 URI=/_bulk
[2020/12/07 17:57:02] [debug] [output:es:es.0] Elasticsearch response
{"took":43,"errors":false,"items":[{"index":{"_index":"fluent-bit","_type":"_doc","_id":"mgvrPXYBDJEY001rpRcG","_version":1,"result":"created","_shards":{"total":2,"successful":2,"failed":0},"_seq_no":31734,"_primary_term":1,"status":201}},{"index":{"_index":"fluent-bit","_type":"_doc","_id":"mwvrPXYBDJEY001rpRcG","_version":1,"result":"created","_shards":{"total":2,"successful":2,"failed":0},"_seq_no":31735,"_primary_term":1,"status":201}},{"index":{"_index":"fluent-bit","_type":"_doc","_id":"nAvrPXYBDJEY001rpRcG","_version":1,"result":"created","_shards":{"total":2,"successful":2,"failed":0},"_seq_no":31736,"_primary_term":1,"status":201}},{"index":{"_index":"fluent-bit","_type":"_doc","_id":"nQvrPXYBDJEY001rpRcG","_version":1,"result":"created","_shards":{"total":2,"successful":2,"failed":0},"_seq_no":31737,"_primary_term":1,"status":201}},{"index":{"_index":"fluent-bit","_type":"_doc","_id":"ngvrPXYBDJEY001rpRcG","_version":1,"result":"created","_shards":{"total":2,"successful":2,"failed":0},"_seq_no":31738,"_primary_term":1,"status":201}},{"index":{"_index":"fluent-bit","_type":"_doc","_id":"nwvrPXYBDJEY001rpRcG","_version":1,"result":"created","_shards":{"total":2,"successful":2,"failed":0},"_seq_no":31739,"_primary_term":1,"status":201}},{"index":{"_index":"fluent-bit","_type":"_doc","_id":"oAvrPXYBDJEY001rpRcG","_version":1,"result":"created","_shards":{"total":2,"successful":2,"failed":0},"_seq_no":31740,"_primary_term":1,"status":201}},{"index":{"_index":"fluent-bit","_type":"_doc","_id":"oQvrPXYBDJEY001rpRcG","_version":1,"result":"created","_shards":{"total":2,"successful":2,"failed":0},"_seq_no":31741,"_primary_term":1,"status":201}},{"index":{"_index":"fluent-bit","_type":"_doc","_id":"ogvrPXYBDJEY001rpRcG","_version":1,"result":"created","_shards":{"total":2,"successful":2,"failed":0},"_seq_no":31742,"_primary_term":1,"status":201}},{"index":{"_index":"fluent-bit","_type":"_doc","_id":"owvrPXYBDJEY001rpRcG","_version":1,"result":"created","_shards":{"total":2,"successful":2,"failed":0},"_seq_no":31743,"_primary_term":1,"status":201}},{"index":{"_index":"fluent-bit","_type":"_doc","_id":"pAvrPXYBDJEY001rpRcG","_version":1,"result":"created","_shards":{"total":2,"successful":2,"failed":0},"_seq_no":31744,"_primary_term":1,"status":201}},{"index":{"_index":"fluent-bit","_type":"_doc","_id":"pQvrPXYBDJEY001rpRcG","_version":1,"result":"created","_shards":{"total":2,"successful":2,"failed":0},"_seq_no":31745,"_primary_term":1,"status":201}},{"index":{"_index":"fluent-bit","_type":"_doc","_id":"pgvrPXYBDJEY001rpRcG","_version":1,"result":"created","_shards":{"total":2,"successful":2,"failed":0},"_seq_no":31746,"_primary_term":1,"status":201}},{"index":{"_index":"fluent-bit","_type":"_doc","_id":"pwvrPXYBDJEY001rpRcG","_version":1,"result":"created","_shards":{"total":2,"successful":2,"failed":0},"_seq_no":31747,"_primary_term":1,"status":201}},{"index":{"_index":"fluent-bit","_type":"_doc","_id":"qAvrPXYBDJEY001rpRcG","_version":1,"result":"created","_shards":{"total":2,"successful":2,"failed":0},"_seq_no":31748,"_primary_term":1,"status":201}},{"index":{"_index":"fluent-bit","_type":"_doc","_id":"qQvrPXYBDJEY001rpRcG","_version":1,"result":"created","_shards":{"total":2,"successful":2,"failed":0},"_seq_no":31749,"_primary_term":1,"status":201}},{"index":{"_index":"fluent-bit","_type":"_doc","_id":"qgvrPXYBDJEY001rpRcG","_version":1,"result":"created","_shards":{"total":2,"successful":2,"failed":0},"_seq_no":31750,"_primary_term":1,"status":201}},{"index":{"_index":"fluent-bit","_type":"_doc","_id":"qwvrPXYBDJEY001rpRcG","_version":1,"result":"created","_shards":{"total":2,"successful":2,"failed":0},"_seq_no":31751,"_primary_term":1,"status":201}},{"index":{"_index":"fluent-bit","_type":"_doc","_id":"rAvrPXYBDJEY001rpRcG","_version":1,"result":"created","_shards":{"total":2,"successful":2,"failed":0},"_seq_no":31752,"_primary_term":1,"status":201}},{"index":{"_index":"fluent-bit","_type":"_doc","_id":"rQvrPX
[2020/12/07 17:57:02] [debug] [upstream] KA connection #32 to ba133372c73b4721a1f45c73ba885114.europe-west1.gcp.cloud.es.io:443 is now available
[2020/12/07 17:57:02] [debug] [task] destroy task=0x7fe014c00210 (task_id=2)
[2020/12/07 17:57:03] [debug] [http_client] not using http_proxy for header
[2020/12/07 17:57:03] [debug] [http_client] header=POST /_bulk HTTP/1.1
Host: ba133372c73b4721a1f45c73ba885114.europe-west1.gcp.cloud.es.io
Content-Length: 2783216
Content-Type: application/x-ndjson
Authorization: Basic ZWxhc3RpYzoydlJjZmZzelNrVXI0eWlxUTJPOEVRWVY=
User-Agent: Fluent-Bit


[2020/12/07 17:57:03] [debug] [http_client] not using http_proxy for header
[2020/12/07 17:57:03] [debug] [http_client] header=POST /_bulk HTTP/1.1
Host: ba133372c73b4721a1f45c73ba885114.europe-west1.gcp.cloud.es.io
Content-Length: 2782857
Content-Type: application/x-ndjson
Authorization: Basic ZWxhc3RpYzoydlJjZmZzelNrVXI0eWlxUTJPOEVRWVY=
User-Agent: Fluent-Bit


[2020/12/07 17:57:04] [debug] [http_client] not using http_proxy for header
[2020/12/07 17:57:04] [debug] [http_client] header=POST /_bulk HTTP/1.1
Host: ba133372c73b4721a1f45c73ba885114.europe-west1.gcp.cloud.es.io
Content-Length: 2781527
Content-Type: application/x-ndjson
Authorization: Basic ZWxhc3RpYzoydlJjZmZzelNrVXI0eWlxUTJPOEVRWVY=
User-Agent: Fluent-Bit
```

### Attached Valgrind output that shows no leaks or memory corruption was found
TBA

ChrsMark commented:
Hey @edsiper @PettitWesley, this one is more or less functional so let me know what you think. Valgrind check is not so healthy but I guess with your review could help with those too.

masci commented on plugins/out_es/es_conf.c:
```diff
@@ -24,10 +24,93 @@
+     /* extract strings */
+    cloud_host_toks = flb_utils_split((const char *)dec_cloud_id, '$', -1);
+    mk_list_foreach(host_head, cloud_host_toks) {
+        items++;
```
nit about style: this index is off-by-one (you're indexing the tokens like 1,2,3), it was a bit unexpected to me (I'd index them 0,1,2)

masci commented on plugins/out_es/es_conf.c:
```diff
@@ -24,10 +24,93 @@
+    mk_list_foreach(host_head, cloud_host_toks) {
+        items++;
+        entry = mk_list_entry(host_head, struct flb_split_entry, _head);
+        if (items == 2) {
```
let's add a comment like `the second token contains the elasticsearch hostname`

masci commented on plugins/out_es/es_conf.c:
```diff
@@ -24,10 +24,93 @@
+          host_hash=entry->value;
+          break;
+        }
+        if (items == 1) {
```
nit (this is my OCD talking 😅): can we put the conditionals in order? Like first evaluate the first token, then the second, then break.

masci commented on plugins/out_es/es_conf.c:
```diff
@@ -24,10 +24,93 @@
+        }
+    }
+    /* format final cloud host */
+    strcpy(cloud_id, host_hash);
```
If `host_hash` is bigger than `cloud_id` it won't end well... Same for subsequent `strcat`s, `cloud_id` must be large enough to contain the result.
I'd return a new buffer to be safe.

ChrsMark replied:
This is correct, but is this possible? `host_hash` and `host_region` are actually items from the split `cloud_id`, so it will never happen.

masci commented on plugins/out_es/es_conf.c:
```diff
@@ -24,10 +24,93 @@
+    struct mk_list *head;
+    struct flb_split_entry *entry;
+
+    toks = flb_utils_split((const char *)cloud_auth, ':', -1);
```
Remember to `flb_utils_split_free(toks)`

masci commented on plugins/out_es/es_conf.c:
```diff
@@ -24,10 +24,93 @@
+        }
+        if (items == 1) {
+          ctx->http_user=entry->value;
+        }
```
same considerations as above for off-by-one and conditionals order :)

masci commented on plugins/out_es/es_conf.c:
```diff
@@ -62,6 +148,21 @@ struct flb_elasticsearch *flb_es_conf_create(struct flb_output_instance *ins,
+    tmp = flb_output_get_property("cloud_id", ins);
+    if (tmp) {
+        cloud_id = (char *) tmp;
+        //cloud_host = extract_cloud_host(ctx, cloud_id);
```
let's remove this if not needed (but it should be needed, right?)

masci commented on plugins/out_es/es_conf.c:
```diff
@@ -73,6 +174,13 @@ struct flb_elasticsearch *flb_es_conf_create(struct flb_output_instance *ins,
+    /* handle cloud_auth */
+    tmp = flb_output_get_property("cloud_auth", ins);
+    if (tmp) {
+        cloud_auth = (char *) tmp;
```
do you need this? I think you can call `set_cloud_credentials(ctx, (char *) tmp);` directly?

masci reviewed (changes requested):
Left few comments, beware this is the first time I look at the codebase :)

ChrsMark commented:
Seems that memory leak issues have been resolved now. `valgrind` output updated within the corresponding [comment](https://github.com/fluent/fluent-bit/pull/2834#issuecomment-740008110).

edsiper commented:
assigned to @fujimotos for review

fujimotos commented:
OK. I spent some time this evening reviewing this PR.

Overall I do not have any objection in feature-wise, but found some issues in the
approach of how to implement it (described in the inline comments). I'm okay with
merging this patch once these nits fixed.

fujimotos commented:
I spent some time this evening reviewing this PR.

Overall I do not have any objection in feature-wise, but found some issues in the
approach of how to implement it (described in the inline comments).

I'm fine with merging this patch once these nits fixed.

ChrsMark commented on plugins/out_es/es_conf.c:
```diff
@@ -24,10 +24,101 @@
+    struct mk_list *host_head = NULL;
+
+    /* keep only part after first ":" */
+    cloud_id_toks = flb_utils_split(cloud_id, ':', 1);
```
Clever approach, thanks! I switch to it, with some modifications since `hash` and `region` should be inverted to shape the final host string. I used `strtok` to split and get only the first part and add in the end of the host-hash string.

ChrsMark commented:
Thanks for the review and suggestions @fujimotos ! I addressed your comments so feel free to have another look when you have the time and let me know!

ChrsMark commented:
Hey @fujimotos, any updates on this one please?

ChrsMark commented:
@edsiper @fujimotos I rebased once more, any plans about moving forward with the review please 🙂 ?

fujimotos commented:
@ChrsMark Can you fetch this commit and check the content?

    https://github.com/fujimotos/fluent-bit/commit/4d4342fa

I applied a few fixes to resolve some issues on your patch.
The highlight is:

 1. Add separate atributes for Elastic Cloud auth to avoid SIGSEGV.

 2. Fix `extract_cloud_host()` and `set_cloud_credentials()` not to
    modify const strings.

 3. Add missing return to the failure path of `extract_cloud_host()`.

I also squashed commits into a single patch and add some information
to the commit message.

If that commit seems good to you, please update this pull request:

```bash
$ git checkout add_cloud_settings
$ git reset --hard origin/master
$ git cherry-pick 4d4342f
$ git push origin add_cloud_settings -f
```

After this, I'm fine with approving this patch.

ChrsMark commented:
Thanks for the edits @fujimotos , I tested back on my end and works fine! I updated the PR so let me know if there is anything else missing.

fujimotos commented:
@edsiper I believe this patch is comittable now.

edsiper commented:
thanks!
